### PR TITLE
implement graceful-stop using a custom AllocationDecider

### PR DESCRIPTION
**Pull request by mfussenegger**
This removes the GracefulStop patch from es/upstream and implements the
graceful stop logic using a custom `DecommissionAllocationDecider`.

Nodes that are being decommissioned are marked in the cluster state as
such and the `DecommissionAllocationDecider` uses these marks to prevent
allocations of shards on those nodes. (Exactly how depends n the
graceful_stop settings)

### CODE REVIEW

File: blackbox/docs/src/crate/process_test.py [hunk 1]
```diff
@@ -215,23 +227,20 @@ def setUp(self):
 
     def test_graceful_stop_none(self):
         """
-        test min_availability: none
+        test `min_availability: none` will stop the node immediately.
+        Causes some shard to become unassigned (no replicas)
```
Comment (seut) [inline blackbox/docs/src/crate/process_test.py]: +1 for the expectation description, would be maybe great to describe this at the other tests also

File: sql/src/main/java/io/crate/cluster/gracefulstop/DecommissionAllocationDecider.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    private Decision canRemainOrAllocate(String nodeId, ShardRouting shardRouting, RoutingAllocation allocation) {
+        if (dataAvailability == DataAvailability.NONE) {
+            return allocation.decision(Decision.YES, NAME, "dataAvailability=none");
+        }
+
+        if (decommissioningNodes.contains(nodeId)) {
+            if (dataAvailability == DataAvailability.PRIMARIES && !shardRouting.primary()) {
+                return allocation.decision(Decision.YES, NAME, "dataAvailability=primaries shard=replica decommissioned=true");
```
Comment (seut) [inline sql/src/main/java/io/crate/cluster/gracefulstop/DecommissionAllocationDecider.java]: looks like the same logic like at canAllocate(), why not remove code from canAllocate() and always forward to this method?

**Comment by mfussenegger** [comment]
todo:
 - update docs regarding allocation.enable settings

File: sql/src/main/java/io/crate/cluster/gracefulstop/DecommissioningService.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                    public void onResponse(ClusterHealthResponse clusterHealthResponse) {
+                        System.exit(0);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        if (forceStop) {
+                            System.exit(0);
+                        }
+                        // TODO: else? should the decommission flag in the settings be removed?
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/cluster/gracefulstop/DecommissioningService.java]: remove the setting to keep old and documented behaviour

File: sql/src/test/java/io/crate/cluster/gracefulstop/DecommissionAllocationDeciderTest.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        DecommissionAllocationDecider allocationDecider =
+            new DecommissionAllocationDecider(settings, mock(NodeSettingsService.class));
+
+        Decision decision = allocationDecider.canAllocate(replicaShard, n1, routingAllocation);
+        assertThat(decision.type(), is(Decision.Type.NO));
+
+        decision = allocationDecider.canRemain(replicaShard, n1, routingAllocation);
+        assertThat(decision.type(), is(Decision.Type.YES));
+    }
```
Comment (seut) [inline sql/src/test/java/io/crate/cluster/gracefulstop/DecommissionAllocationDeciderTest.java]: pls add a test for full availability

**Comment by seut** [comment]
👍 ack

**Pull request merged**